File: app/src/main/java/ru/vladislav_akulinin/mychat_version_2/Fragments/ChatsFragment.java
package ru.vladislav_akulinin.mychat_version_2.Fragments;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import ru.vladislav_akulinin.mychat_version_2.Adapter.UserAdapter;
import ru.vladislav_akulinin.mychat_version_2.Model.Chat;
import ru.vladislav_akulinin.mychat_version_2.Model.Chatlist;
import ru.vladislav_akulinin.mychat_version_2.Model.User;
import ru.vladislav_akulinin.mychat_version_2.R;

public class ChatsFragment extends Fragment {

    private RecyclerView recyclerView;

    private UserAdapter userAdapter;
    private List<User> mUsers;

    FirebaseUser fuser;
    DatabaseReference reference;

    private List<String> usersList; //оптимизируем код
//    private List<Chatlist> usersList;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_chats, container, false);

        recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        fuser = FirebaseAuth.getInstance().getCurrentUser();

        usersList = new ArrayList<>();

        //заменяем данный код (для оптимизации)
        reference = FirebaseDatabase.getInstance().getReference("Chats");
        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                usersList.clear();

                for(DataSnapshot snapshot : dataSnapshot.getChildren()){
                    Chat chat = snapshot.getValue(Chat.class);

                    if(chat.getSender().equals(fuser.getUid())){
                        usersList.add(chat.getReceiver());
                    }
                    if(chat.getReceiver().equals(fuser.getUid())){
                        usersList.add(chat.getSender());
                    }
                }

                readChats();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

//        reference = FirebaseDatabase.getInstance().getReference("Chatlist").child(fuser.getUid());
//        reference.addValueEventListener(new ValueEventListener() {
//            @Override
//            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
//                usersList.clear();
//                for(DataSnapshot snapshot : dataSnapshot.getChildren()){
//                    Chatlist chatlist = snapshot.getValue(Chatlist.class);
//                    usersList.add(chatlist);
//                }
//
//                chatList();
//            }
//
//            @Override
//            public void onCancelled(@NonNull DatabaseError databaseError) {
//
//            }
//        });

        return view;
    }

//    private void chatList() {
//
//        mUsers = new ArrayList<>();
//        reference = FirebaseDatabase.getInstance().getReference("User");
//        reference.addValueEventListener(new ValueEventListener() {
//            @Override
//            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
//                mUsers.clear();
//                for(DataSnapshot snapshot : dataSnapshot.getChildren()){
//                    User user = snapshot.getValue(User.class);
//                    for(Chatlist chatlist : usersList){
//                        if(user.getId().equals(chatlist.getId())){
//                            mUsers.add(user);
//                        }
//                    }
//                }
//                userAdapter = new UserAdapter(getContext(), mUsers,true);
//                recyclerView.setAdapter(userAdapter);
//            }
//
//            @Override
//            public void onCancelled(@NonNull DatabaseError databaseError) {
//
//            }
//        });
//
//    }

//    заменяем данный код (для оптимизации)
    private void readChats(){ //читать сообщения
        mUsers = new ArrayList<>();

        reference = FirebaseDatabase.getInstance().getReference("User");

        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                mUsers.clear();

                for (DataSnapshot snapshot : dataSnapshot.getChildren()){
                    User user = snapshot.getValue(User.class);

                    //показывать 1 пользователя из чатов
                    for(String id : usersList){
                        if (user.getId().equals(id)){
                            if(mUsers.size() != 0){
                                for(User user1 : mUsers){
                                    if(!user.getId().equals(user1.getId())){
                                        mUsers.add(user);
                                    }
                                }
                            } else {
                                mUsers.add(user);
                            }
                        }
                    }
                }
                                                                        //try - для статуса
                userAdapter = new UserAdapter(getContext(), mUsers, true);
                recyclerView.setAdapter(userAdapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }
    
}

File: app/src/main/java/ru/vladislav_akulinin/mychat_version_2/chat/ChatsFragment.kt
package ru.vladislav_akulinin.mychat_version_2.chat

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import kotlinx.android.synthetic.main.fragment_chats.view.*
import ru.vladislav_akulinin.mychat_version_2.R
import ru.vladislav_akulinin.mychat_version_2.adapter.UserAdapter
import ru.vladislav_akulinin.mychat_version_2.model.UserModel
import com.google.firebase.database.FirebaseDatabase


class ChatsFragment : Fragment() {

    private lateinit var userAdapter: UserAdapter
    private var userModelList: MutableList<UserModel>? = null

    private val firebaseUser = FirebaseAuth.getInstance().currentUser
    private val firebaseDatabase: Query = FirebaseDatabase.getInstance().reference


    @SuppressLint("PrivateResource")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.fragment_chats, container, false)

        val layoutManager = LinearLayoutManager(context)
        view.recycler_view.layoutManager = layoutManager

        val dividerItemDecoration = DividerItemDecoration(view.recycler_view.context, layoutManager.orientation)
        view.recycler_view.addItemDecoration(dividerItemDecoration)

        val fabCreateNewChar: View = view.findViewById(R.id.fab_create_chat)
        fabCreateNewChar.setOnClickListener {
            fragmentManager!!
                    .beginTransaction()
                    .setCustomAnimations(R.anim.design_bottom_sheet_slide_in, R.anim.design_bottom_sheet_slide_out)
                    .replace(R.id.container, AddChatFragment())
                    .addToBackStack(null)
                    .commit()
        }

        return view
    }

    private fun newChat(fabCreateNewChar: View, view: View) {


//        fabCreateNewChar.setOnClickListener {
//            val input = view.findViewById(R.id.input) as EditText
//
//            firebaseDatabase
//                    .child("Chats")
//                    .push()
//                    .setValue(ChatModel(input.text.toString(),
//                            FirebaseAuth.getInstance()
//                                    .currentUser!!
//                                    .displayName)
//                    )
//
//            input.setText("")
//        }
    }

//    private fun readChats() {
//        userModelList = ArrayList<User>()
//
//        reference = FirebaseDatabase.getInstance().getReference("UserNew")
//
//        reference.addValueEventListener(object : ValueEventListener {
//            override fun onDataChange(dataSnapshot: DataSnapshot) {
//                mUserJavas.clear()
//
//                for (snapshot in dataSnapshot.children) {
//                    val userJava = snapshot.getValue(UserJava::class.java)
//
//                    //показывать 1 пользователя из чатов
//                    for (id in usersList) {
//                        if (userJava!!.id == id) {
//                            if (mUserJavas.isEmpty()) {
//                                for (userJava1 in mUserJavas) {
//                                    if (userJava.id != userJava1.getId()) {
//                                        mUserJavas.add(userJava)
//                                    }
//                                }
//                            } else {
//                                mUserJavas.add(userJava)
//                            }
//                        }
//                    }
//                }
//                userAdapterJava = UserAdapterJava(context, mUserJavas, true)
//                recyclerView.setAdapter(userAdapterJava)
//            }
//
//            override fun onCancelled(databaseError: DatabaseError) {
//
//            }
//        })
//    }
}